### PR TITLE
session stats now subscribe to session events

### LINKED ISSUE #621: Session duration is reset after reconnecting in dev-net
Connection for mobile is stable in dev-net, but each time reconnect happens - session duration (received from statistics endpoint) is reset to 0.

### PR DESCRIPTION
**Pull request by vkuznecovas**
Connection manager now emits session events, which the session storage, statistics tracker/reporter now subscribe to. This ensures that internal connection state does not affect the statistics and in return, does not reset the session duration.

Closes #621

### CODE REVIEW

File: consumer/statistics/tracker.go [hunk 1]
```diff
@@ -69,17 +70,36 @@ func (sst *SessionStatisticsTracker) markSessionEnd() {
 	sst.sessionStart = nil
 }
 
+// calcStatDiff takes in the old and the new values of statistics, returns the calculated delta
+func (sst *SessionStatisticsTracker) calcStatDiff(old, new uint64) (res uint64) {
```
Comment (tadovas) [inline consumer/statistics/tracker.go]: It's just a function - no need of receiver

File: consumer/statistics/tracker.go [hunk 2]
```diff
@@ -69,17 +70,36 @@ func (sst *SessionStatisticsTracker) markSessionEnd() {
...
+	if old > new {
+		return new
+	}
+	return new - old
+}
+
+// GetStatisticsDiff calculates the difference in bytes between the old stats and new
+func (sst *SessionStatisticsTracker) GetStatisticsDiff(old, new consumer.SessionStatistics) consumer.SessionStatistics {
```
Comment (tadovas) [inline consumer/statistics/tracker.go]: Same here - no need of receiver

File: consumer/session/session_storage.go [hunk 3]
```diff
@@ -66,17 +66,17 @@ func (repo *Storage) GetAll() ([]History, error) {
...
-func (repo *Storage) ConsumeStateEvent(stateEvent connection.StateEvent) {
-	switch stateEvent.State {
-	case connection.Disconnecting:
-		repo.handleDisconnectingEvent(stateEvent.SessionInfo.SessionID)
-	case connection.Connected:
-		repo.handleConnectedEvent(stateEvent.SessionInfo)
+// ConsumeSessionEvent consumes the session state change events
+func (repo *Storage) ConsumeSessionEvent(sessionEvent connection.SessionEvent) {
+	switch sessionEvent.Status {
+	case connection.SessionStatusEnded:
```
Comment (zolia) [inline consumer/session/session_storage.go]:
```suggestion
	case connection.SessionEndedStatus:
```
is more intuitive.

File: consumer/statistics/tracker.go [hunk 4]
```diff
@@ -69,17 +70,36 @@ func (sst *SessionStatisticsTracker) markSessionEnd() {
...
+		BytesSent:     sst.calcStatDiff(old.BytesSent, new.BytesSent),
+		BytesReceived: sst.calcStatDiff(old.BytesReceived, new.BytesReceived),
+	}
+}
+
 // ConsumeStatisticsEvent handles the connection statistics changes
 func (sst *SessionStatisticsTracker) ConsumeStatisticsEvent(stats consumer.SessionStatistics) {
-	sst.sessionStats = stats
+	diff := sst.GetStatisticsDiff(sst.lastStats, stats)
+	sst.sessionStats.BytesReceived += diff.BytesReceived
```
Comment (tadovas) [inline consumer/statistics/tracker.go]: If you are making a function to calc diff, maybe it's worth making a function newStats := AddUp(sessionStats, diff)

Comment (tadovas) [reply consumer/statistics/tracker.go]: Those diff and add methods can even be part of SessionStatistics since we are doing calculations with them anyway

**Pull request merged**